### PR TITLE
implementing #10

### LINKED ISSUE #10: Update config.yaml
> That file represents the state of the access as configured in GitHub at the time I created it. If it does not represent the community, or the decisions made therein, please file an issue to document the changes required.

_Originally posted by @ryjones in https://github.com/open-quantum-safe/tsc/issues/6#issuecomment-2027251771_
            
As requested: Ensure all (sub-)project roles of contributor- and maintainership are properly represented.

At the least implement https://github.com/open-quantum-safe/tsc/issues/7#issuecomment-2013658258

Then I think @thomwiggers no longer maintains liboqs-rust, right?
@vsoftco in turn should be maintainer of all language wrappers
@baentsch is maintainer of oqs-provider and liboqs; @thb-sb: Would you want [to commit being maintainer in either or just one project](https://github.com/open-quantum-safe/oqs-provider/blob/main/GOVERNANCE.md#committers), too?
@dstebila is no maintainer of oqs-provider but surely of liboqs.
All other projects don't have maintainers, AFAIK (which is a problem and should be tackled -- as per https://github.com/open-quantum-safe/tsc/issues/2

Anyone else, please chime in if I forgot about you(r commitments).

### PR DESCRIPTION
**Pull request by baentsch**
Fixes #10 as per https://github.com/open-quantum-safe/tsc/issues/10#issuecomment-2339400591 

Draft only pending TSC agreement

### CODE REVIEW

**Comment by baentsch** [comment]
As I didn't hear adamant opposition to this proposal in the TSC meeting, moving to Ready for Review

File: .github/pull_request_template.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+<!-- Does this PR resolve any issue?  If so, please reference it using automatic-closing keywords like "Fixes #123." -->
+
+<!-- As changes to the file `config.yaml` are particularly sensitive because they change GH permissions throughout the project, the following rules apply to PRs affecting this file -->
+
+Unconditionally, changes to `config.yaml` must
+- [ ] be approved by 2 members of the OQS TSC
+- [ ] not violate permissions documented in GOVERNANCE.md files for sub projects where such files exist
+
+The following goals apply to changes to the file `config.yaml` with exceptions possible, as long as the rationale for the excption is documented by comments in the file:
```
Comment (SWilson4) [inline .github/pull_request_template.md:11]:
```suggestion
The following goals apply to changes to the file `config.yaml` with exceptions possible, as long as the rationale for the exception is documented by comments in the file:
```

Very minor typo fix.

**Review by dstebila — APPROVED** [review]

**Comment by baentsch** [comment]
As expected, I just don't understand this file structure/rules. @SWilson4 Would you mind taking a second look/improve on this as per your kind offer? @open-quantum-safe/tsc Do _not_ review (I don't seem to have permission to revert to Draft....). Sorry for the message spam.

**Comment by ryjones** [comment]
@baentsch teams cannot contain teams; that's the error

**Comment by baentsch** [comment]
> @baentsch teams cannot contain teams; that's the error

Thanks for the hint, @ryjones  So this tool has no way to express hierarchies? Every individual team (membership) must be manually set and curated? Then what's the advantage of this tool? I'd be faster clicking through GH UIs than tediously editing things this way. Would there be an alternative way to split/parcel out responsibility of the whole file somehow instead? Say OQS-global stuff here, liboqs permissions there, and so on for/within each sub project?

**Comment by baentsch** [comment]
> One advantage is that there is a public audit log of changes, and as a "dsc" file (desired state configuration), if there is a change, it will be reverted.

OK, the "revert inadvertent changes" feature sounds sensible. But does this offset months of running with a largely wrong and even insecure configuration, or making people waste hours and days having to work around problems created this way? Is it possible process is ruling safety, security, economics and common sense here? Do you know of any other (better, safer) way to achieve this, @ryjones ?

Back to my question: We have to really add individual GH ids repeatedly to each and every team?

If so, I have strong doubts that this can be maintained in the long run in a safe and reliable manner; particularly if more people would join the project and gaining (or losing) "cred" and thus permissions, this becomes totally unmanageable. Already the current state of things kind-of confirms this: For example, people completely external to the project have been given admin access to `oqsprovider` while that access has been withheld from its maintainer also for months.

I assume you are most likely just following orders -- so could you please point to some decision document/meeting that agreed using this tool and procedure, @ryjones so we could discuss better, more safe and more scalable alternatives going forward in that venue? Was that the TAC? The only alternative I see is PQCA hires a dedicated security admin for the project, intimately familiar with who does what -- but honestly, I'd prefer spending money on developers and testers. Any objections to that statement @dstebila ?

**Comment by SWilson4** [comment]
For now, I'll take a stab at bringing the file in line with GOVERNANCE documents and discussion in issue #10.

Just wanted to check with @thb-sb: if I recall, you said in the most recent meeting that you were no longer working at Sandbox. Does that mean that we should now be using a different GitHub handle for you in the config?

**Review by ryjones — APPROVED** [review]

**Comment by SWilson4** [comment]
OK, I've taken a stab at redoing this configuration. Please take a look @open-quantum-safe/tsc.

Here's a plain-English explanation of what I tried to do with the config file. If people think this analysis is valuable, I can also write it up and include it as a Markdown file for this repo.

Please note that the distinction I make between "organization-level" and "repository-level" teams is one of organization. There is **no functional difference** in how GitHub views the teams.

## Organization-level teams

These are teams which are not tied to a specific subproject and will generally have permissions across all repos in the organization.

### tsc
The OQS Technical Steering Committee. This team should have `read` access to all repositories and `write` access to the tsc repo.

### oqs-admins
For general GitHub administrators. Think of this as the "sysadmin" team. This team should have `admin` access to all repositories.

### bots
For accounts that we use to automate actions (e.g., uploading Docker images or triggering CI jobs across projects). This team should have `write` access to all repositories. (This is the minimum permission level required to trigger workflows.)

### triage
For people who need to do issue management but are not Contributors. This team should have `triage` access to all repositories.

### trail-of-bits
Created solely to allow Trail of Bits members access to a project board tracking issues from the code audit. We can probably drop this team once the final report is released. This team doesn't need any permissions; it just pulls the ToB people into the organization.

## Repository-level teams

Each technical subproject has the following associated teams:
- Maintainers (with `admin` access)
- Release Managers (with `maintain` access) --- if the subproject does releases
- Committers (with `write` access)
- Contributors (with `write` access)
 
**Wherever possible, I followed a GOVERNANCE.md file when defining a team.** When a GOVERNANCE.md file was not available or did not specify a team, I first made a best effort to encode "[folklore](https://en.wikipedia.org/wiki/Mathematical_folklore)" (for example, setting @vsoftco as the maintainer of liboqs-go, liboqs-cpp, and liboqs-python). To handle the cases when Maintainers / Committers / etc. were not clearly specified or agreed on by unwritten consensus, I created what I consider to be sane default teams: **oqs-maintainers**, **oqs-release-managers**, **oqs-committers**, and **oqs-contributors**.

### oqs-maintainers
Consists of TSC members who are also maintainers of an active subproject.

### oqs-release managers
Consists of the above team plus "full-time" OQSers (i.e., myself and @praveksharma).

### oqs-committers
Consists of the previous "core" GitHub team.

### oqs-contributors
People who have made significant code contributions and/or are listed in a CODEOWNERS file for some repository. Please take a look at this team in particular and see if I'm missing anybody.

## Corner cases

### openssl
This repo is inactive and probably should be archived. I gave it what I considered to be the bare minimum set of teams (**oqs-admin** and **tsc**).

### liboqs-cupqc
This repo is private and is doing its own thing.

### minute-takers
This team consists of people who are responsible for sharing minutes from OQS status calls (currently, me). It has write access to the tsc repo so that minutes can be uploaded without code review.

**Comment by SWilson4** [comment]
@baentsch in particular, please take a look and let me know if this faithfully captures what you had in mind with in the discussion in #10.

**Comment by SWilson4** [comment]
>     * The "release managers" teams do not constitute the same people in each case: Intention or proof point that [this tool is a danger to the safe and coherent maintenance of code access](https://github.com/open-quantum-safe/tsc/pull/77#issuecomment-2361436662)? Errors, particularly to the "equal treatment rule" of people and sub projects, are practically impossible to spot.
 
I tried to construct the subproject-specific "release managers" teams based on who'd done past releases. The only subprojects without the default list (oqs-release-managers) are
- OpenSSH, whose list includes Gerardo and excludes myself;
- liboqs-cpp / liboqs-go / liboqs-python, whose list consists solely of Vlad; and
- liboqs-rust, whose list consists solely of Thom.

If it's preferable to use the default list across all projects, I can make that switch easily, but I thought the exceptions made sense (especially for liboqs-rust).

>     * A global "oqs-contributors" list is absolutely not to my liking (and in contradiction to what is documented in the PR template/documentation): This gives full project write access to all sub projects even if only a tiny slice of code is "owner-tagged" in a single sub project: IMO this clearly runs counter to the least privilege concept. That said, the big problem in this space for oqs-provider seems gone: Thanks!
 
I agree absolutely. Ideally we would draw this information in GOVERNANCE files (I think?). In the absence of that I opted for the global list as choosing an ad-hoc list of contributors for each subproject seemed like an ordeal. :/

>     * This does not address key person risk in any coherent manner: There is still _one_ admin and _one_ maintainer in most teams; the one big thing that improved there is that maintainers now also have admin rights again returning control to all sub community leaders. It's too few of them, though: In practically all cases, I'd think both @dstebila and @baentsch should be listed (and one can argue whether that's enough given we both do other things more and more often).
 
I think the usage of this config file actually does help to address key person risk—someone without any admin privileges whatsoever (such as myself) can propose config changes (e.g., upgrading somebody admin privileges), and they will be made upon merge. As long as there are enough people who can approve PRs in this repo (i.e., the TSC), then I believe that we could very quickly delegate emergency admin powers to somebody in the event that a key person (say @ryjones as the universal admin) goes AWOL. We would just need to make a PR adding one line to this file (to add somebody to the oqs-admins team) and get it merged.

(I should ask @ryjones to fact-check the above statement just in case I'm misunderstanding the power of this file)

>     * On the positive side, your comments pointing to GOVERNANCE.md files and pointing out the many TODOs are excellent @SWilson4

Thank you! :)

**Comment by ryjones** [comment]
You might notice @thelinuxfoundation is also an enterprise owner, and therefore an org owner. There are about a dozen people that can use that account to bootstrap into an org with no other owners; they are all Linux Foundation employees.

**Comment by ryjones** [comment]
@SWilson4 @baentsch I made an explicit group to list the org owners.

**Comment by dstebila** [comment]
>     * A global "oqs-contributors" list is absolutely not to my liking (and in contradiction to what is documented in the PR template/documentation): This gives full project write access to all sub projects even if only a tiny slice of code is "owner-tagged" in a single sub project: IMO this clearly runs counter to the least privilege concept. That said, the big problem in this space for oqs-provider seems gone: Thanks!

Yes, this also strikes me as casting the net for people with write access a bit too wide.  In some of our [GOVERNANCE.md files](https://github.com/open-quantum-safe/liboqs/blob/main/GOVERNANCE.md) we distinguish between committers as having write access and contributors as not having write access, and I think that's a good distinction to maintain.  

>     * This does not address key person risk in any coherent manner: There is still _one_ admin and _one_ maintainer in most teams; the one big thing that improved there is that maintainers now also have admin rights again returning control to all sub community leaders. It's too few of them, though: In practically all cases, I'd think both @dstebila and @baentsch should be listed (and one can argue whether that's enough given we both do other things more and more often).

I'm less concerned about key person risk from a "lock-out" perspective -- the LF ownership of the org will ensure that someone continues to have access to the repository.  I think it is a legitimate concern that we don't have enough people in certain roles (e.g., not enough people taking on maintainer roles), but that's not something that can be fixed in a config file, it's about building and strengthening the community.

File: config.yaml [hunk 2]
```diff
@@ -1,281 +1,428 @@
...
-    claucece: write
-    pi-314159: write
+
+- name: liboqs
   teams:
-    boringssl-maintainers: write
+    oqs-admins: admin
+    liboqs-maintainers: admin
+    oqs-release-managers: maintain
+    liboqs-committers: write
```
Comment (dstebila) [inline config.yaml:241]: It seems like the permissions provided to `<projectname>-committers` is inconsistent. In liboqs, the `liboqs-committers` team gets `write`, but below in oqs-providers, the `oqs-providers-committers` team gets `maintain`.  Is that intentional?

Comment (SWilson4) [reply config.yaml:241]: That's an oversight on my part. I intended to give all Committers teams write access.

**Comment by SWilson4** [comment]
OK, I've made another iteration to (hopefully) address the least-privilege concerns. For subprojects with CODEOWNERS files (currently only liboqs and oqs-provider), I've created a `<subproject>-codeowners` team with write access consisting of the people in the CODEOWNERS file. I've downgraded the Contributors team access to `triage` across all subprojects and merged it with the "triage" team.

File: config.yaml (hunk 2 above)
Comment (SWilson4) [reply config.yaml:258]: Fixed in [236ba1c](https://github.com/open-quantum-safe/tsc/pull/77/commits/236ba1ce1b575dc1c73a8fea96dd0ef46b6a5762).

**Review by dstebila — APPROVED** [review]

**Comment by dstebila** [comment]
I see that @ryjones just merged this, and that made me realize we should make sure we understand what one of the phrases means in the new requirements, specifically: "be approved by 2 members of the OQS TSC".  Michael and I are both TSC members; he authored (implying he approves of it) and I gave an approving code review; does that suffice for this purpose?  Or are we interpreting this statement to mean that there should be 2 approvals by TSC members on top of any TSC members who are authors of the PR?

**Comment by baentsch** [comment]
> are we interpreting this statement to mean that there should be 2 approvals by TSC members on top of any TSC members who are authors of the PR?

An author approving of his work leads the concept of review ad absurdum.

Also the author's intention with this verbiage was to stress the fact that GH permissions control has more impact on the security of the overall system's code than the number of reviews on a single PR (where `liboqs` for example already has the 6-eyes principle).

File: config.yaml [hunk 3]
```diff
@@ -1,281 +1,432 @@
...
+
+# liboqs-rust Maintainers
+# TODO: provide "source of truth"
+- name: liboqs-rust-maintainers
   maintainers:
-  - crockeea
+  - thomwiggers
+# liboqs-rust Release Managers
+# TODO: provide "source of truth"
+- name: liboqs-rust-release-managers
```
Comment (thomwiggers) [inline config.yaml:225]: By renaming this team, the Crates.io ownership got invalidated. Fortunately, I'm still a direct owner of the `oqs` crates on `crates.io`.

See https://crates.io/teams/github:open-quantum-safe:rust

Comment (dstebila) [reply config.yaml:225]: Thanks for letting us know, Thom. Is it easy to change the team reference in crates.io?  If not we can change the team name back here.  Either way @SWilson4 can we create a comment in the file noting that the team name is connected to permissions on crates.io?

Comment (SWilson4) [reply config.yaml:225]: Yes—I'll put up a PR to do that. If necessary I'll also change the team name back to "rust"; just let me know @thomwiggers.

Comment (SWilson4) [reply config.yaml:225]: See https://github.com/open-quantum-safe/tsc/pull/85.

**Pull request merged**